RecoLocalTracker/SiPixelRecHits/src/PixelCPEFast.cc

```diff
@@ -0,0 +1,594 @@
+  xerr_endcap_def_ = 0.0020;
+  yerr_endcap_ = {0.00210};
+  yerr_endcap_def_ = 0.00075;
+
+  fillParamsForGpu();
+
+  cpuData_ = {
+      &m_commonParamsGPU,
+      m_detParamsGPU.data(),
+      &m_layerGeometry,
+      &m_averageGeometry,
+  };
+}
+
+const pixelCPEforGPU::ParamsOnGPU* PixelCPEFast::getGPUProductAsync(cudaStream_t cudaStream) const {
+  const auto& data = gpuData_.dataForCurrentDeviceAsync(cudaStream, [this](GPUData& data, cudaStream_t stream) {
+    // and now copy to device...
+    cudaCheck(cudaMalloc((void**)&data.h_paramsOnGPU.m_commonParams, sizeof(pixelCPEforGPU::CommonParams)));
+    cudaCheck(cudaMalloc((void**)&data.h_paramsOnGPU.m_detParams,
+                         this->m_detParamsGPU.size() * sizeof(pixelCPEforGPU::DetParams)));
+    cudaCheck(cudaMalloc((void**)&data.h_paramsOnGPU.m_averageGeometry, sizeof(pixelCPEforGPU::AverageGeometry)));
+    cudaCheck(cudaMalloc((void**)&data.h_paramsOnGPU.m_layerGeometry, sizeof(pixelCPEforGPU::LayerGeometry)));
+    cudaCheck(cudaMalloc((void**)&data.d_paramsOnGPU, sizeof(pixelCPEforGPU::ParamsOnGPU)));
+
+    cudaCheck(cudaMemcpyAsync(
+        data.d_paramsOnGPU, &data.h_paramsOnGPU, sizeof(pixelCPEforGPU::ParamsOnGPU), cudaMemcpyDefault, stream));
+    cudaCheck(cudaMemcpyAsync((void*)data.h_paramsOnGPU.m_commonParams,
+                              &this->m_commonParamsGPU,
+                              sizeof(pixelCPEforGPU::CommonParams),
+                              cudaMemcpyDefault,
+                              stream));
+    cudaCheck(cudaMemcpyAsync((void*)data.h_paramsOnGPU.m_averageGeometry,
+                              &this->m_averageGeometry,
+                              sizeof(pixelCPEforGPU::AverageGeometry),
+                              cudaMemcpyDefault,
+                              stream));
+    cudaCheck(cudaMemcpyAsync((void*)data.h_paramsOnGPU.m_layerGeometry,
+                              &this->m_layerGeometry,
+                              sizeof(pixelCPEforGPU::LayerGeometry),
+                              cudaMemcpyDefault,
+                              stream));
+    cudaCheck(cudaMemcpyAsync((void*)data.h_paramsOnGPU.m_detParams,
+                              this->m_detParamsGPU.data(),
+                              this->m_detParamsGPU.size() * sizeof(pixelCPEforGPU::DetParams),
+                              cudaMemcpyDefault,
+                              stream));
+  });
+  return data.d_paramsOnGPU;
+}
+
+void PixelCPEFast::fillParamsForGpu() {
+  m_commonParamsGPU.theThicknessB = m_DetParams.front().theThickness;
+  m_commonParamsGPU.theThicknessE = m_DetParams.back().theThickness;
+  m_commonParamsGPU.thePitchX = m_DetParams[0].thePitchX;
+  m_commonParamsGPU.thePitchY = m_DetParams[0].thePitchY;
+
+  // std::cout << "pitch & thickness " <<  m_commonParamsGPU.thePitchX << ' ' << m_commonParamsGPU.thePitchY << "  " << m_commonParamsGPU.theThicknessB << ' ' << m_commonParamsGPU.theThicknessE << std::endl;
```

Comment:
change to logDebug and uncomment?

Reply:
Done in https://github.com/cms-patatrack/cmssw/pull/593/commits/9dc393168636f3c4e48986f7739c554f75cdd1c1 .